### PR TITLE
Add TfidfVectorizer & 20 newsgroups replication script

### PR DESCRIPTION
**Pull request by x65han**
- added new TfidfVectorizer class
- added bin/get-20newsgroups-data.sh & scripts/20newsgroups-replication.py (with a f1 score of `0.8099420314961255`)
- added `docs/20newsgroups.md` with replication instructions

Replication steps from pyserini root folder:
```
sh bin/get-20newsgroups-data.sh
python scripts/20newsgroups-replication.py
```

### CODE REVIEW

File: pyserini/vectorizer.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import math
+import numpy as np
+from typing import List
+from scipy.sparse import csr_matrix
```
Comment (x65han) [inline pyserini/vectorizer.py:4]: Do we need to update [requirements.txt](https://github.com/castorini/pyserini/blob/master/requirements.txt) to include `numpy` and `scipy`?

Comment (lintool) [reply pyserini/vectorizer.py:4]: yes

File: bin/20newsgroup-replication.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+sys.path.insert(0, "./")
+
+from pyserini.vectorizer import TfidfVectorizer
+import os
+from sklearn.naive_bayes import MultinomialNB
+from sklearn import metrics
+
+
+# Build labels
+target_names = ['alt.atheism', 'comp.graphics', 'comp.os.ms-windows.misc', 'comp.sys.ibm.pc.hardware', 'comp.sys.mac.hardware', 'comp.windows.x', 'misc.forsale', 'rec.autos', 'rec.motorcycles', 'rec.sport.baseball',
```
Comment (lintool) [inline bin/20newsgroup-replication.py:11]: lines too long

File: bin/20newsgroup-replication.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    for root, _, files in os.walk(path, topdown=False):
+        for doc_id in files:
+            docs.append(doc_id)
+            category = root.split('/')[-1]
+            targets.append(target_to_index[category])
+
+    return docs, targets
+
+
+train_docs, train_labels = get_info('./20-newsgroup/20news-bydate-train/')
```
Comment (lintool) [inline bin/20newsgroup-replication.py:31]: add a main?

File: bin/20newsgroup-replication.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import sys
```
Comment (lintool) [inline bin/20newsgroup-replication.py:1]: hrm, I'm wondering if we should have a separate `scripts/` directory and save `bin/` just for simple shell scripts?

File: pyserini/vectorizer.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        self.index_utils = pyutils.IndexReaderUtils(lucene_index_path)
+
+        # get num_docs
+        self.searcher = pysearch.SimpleSearcher(lucene_index_path)
+        self.num_docs: int = self.searcher.num_docs
+
+        # pre-processing
+        self.compute_idf_and_build_vocab()
+
+    def compute_idf_and_build_vocab(self):
```
Comment (lintool) [inline pyserini/vectorizer.py:21]: shouldn't this just be part of init?

File: pyserini/vectorizer.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import math
+import numpy as np
+from typing import List
+from scipy.sparse import csr_matrix
+from pyserini.search import pysearch
+from pyserini.index import pyutils
+
+
+class TfidfVectorizer:
+    def __init__(self, lucene_index_path: str, min_df: int = 0) -> None:
```
Comment (lintool) [inline pyserini/vectorizer.py:10]: set default to one - i.e., discard singleton terms by default.

**Comment by lintool** [comment]
You're going to add documentation in `docs/` right?

File: bin/20newsgroup-replication.py (hunk 4 above)
Comment (x65han) [reply bin/20newsgroup-replication.py:1]: I can do that.

File: pyserini/vectorizer.py (hunk 5 above)
Comment (x65han) [reply pyserini/vectorizer.py:21]: I thought that maybe we want to refactor this for later. 

merging with the constructor for now and will refactor later if there's the need.

File: pyserini/vectorizer.py (hunk 1 above)
Comment (x65han) [reply pyserini/vectorizer.py:4]: added

**Comment by x65han** [comment]
@lintool ready for round 2 review. Thanks

File: scripts/get-20newsgroup-data.sh [hunk 7]
```diff
@@ -0,0 +1,8 @@
+FOLDER=20-newsgroup
```
Comment (lintool) [inline scripts/get-20newsgroup-data.sh:1]: hrm, I was thinking of `bin/` for shell scripts and `scripts/` for end-to-end Python scripts... but open to discussion?

Comment (x65han) [reply scripts/get-20newsgroup-data.sh:1]: Ok, I'm fine with that.


I will make the following changes:
- bin/get-20newsgroup-data.sh
- scripts/generate_trec_covid_round1_OSHU_RUN2.py
- scripts/20newsgroup-replication.py

Comment (x65han) [reply scripts/get-20newsgroup-data.sh]: Done

**Pull request merged**